### PR TITLE
Make sure kernel segmentation pass merges all valid pieces, and deterministically

### PR DESCRIPTION
**Pull request by shmsong**
- Add a final merging pass to make sure the kernel segmenter merges all valid node pairs.
- Changed mergeNodes implementation to give deterministic merging result.
- Remove unnecessary assertion on group infer input.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: Is this safe? Is it guaranteed to have `key` in the map?

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: I'd suggest using `at` unless the existence of a given key is extremely clear.

Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: Yes. Operator [] just creates an empty pointer if not matched, and the emptiness is checked in the following lines.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: You're right. I forgot unordered_map automatically creates a default value if key not found.

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp:1086]: Can you please comment here and describe why this `finalMerge` is necessary? A simple example case would be great.

Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp:1086]: Thanks for pointing this out. Added comments in fusion_segmenter.h and a motivating simpler example as a test `FusionDAGMerging_CUDA`.

Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp:1086]: While the example is quite synthetical, I do expect we hit quite a few similar cases where the algorithm on the paper cannot propose a valid merge. One of the main cause I observed now is that we have additional constraints on what kinds of nodes can merge and the original algorithm is a hierarchical/iterative one. So it can be terminated early by our constraints.

File: test/cpp/jit/test_gpu.cpp
Comment (naoyam) [inline test/cpp/jit/test_gpu.cpp:13433]: I'm confused with this comment. Does each number mean a tensor? E.g., `{0}` means `{tv0}`? If so, what about `tv6` and `tv7`?

**Comment by naoyam** [comment]
@shmsong I merged #724 and this and still got a root-mapping error with the EnsureMerging test. There seems a different issue with the fusion. Since it doesn't seem non-trivial, I created another issue with the EnsureMerging test (#725). I think we should remove it from this PR and merge it to the devel branch.

File: test/cpp/jit/test_gpu.cpp
Comment (csarofeen) [reply test/cpp/jit/test_gpu.cpp:13433]: The expression number, which is in the order of `fusion::exprs()`

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: Can you make this function name a bit more specific. insert is such a common container function to see something so simple as `insert(to, g)` just seems a bit off. `insertBackwardReachable` is verbose to the point it may seem unnecessary, but it make the point immediately clear that it's a function inserting something associated with the class we're looking at. I understand what's going on, but there's a lot of very generic sounding functions:
`getSet`
`insert`
`work_list`
`to_update`
I'm not recommending going and changing all of these, insert was just the one that threw me off the most. However, writing code should be like writing a document, and I always encourage names that can get the point across quickly. For example one of Carilli's function name: `currentStreamCaptureStatusMayInitCtx` That's quite verbose, but also leaves little to the imagination.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: This function also seems like it provides very minimal abstraction, may want to leave it out alltogether.

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: Nit: I've typically been sticking with with `to_visit` and `visited`. Just makes them seem more related with that language.

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: Stick with how you would write. For group in groups. For consumer_edge in consumer_edges. It makes it much easier to read when using words instead of the first letter. If it starts getting really verbose then we can shorten them, but we don't need single letter variables, it's not code golf ;-)

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: `all_consumers_of_consumer` otherwise this also gets a bit confusing.

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: reachability seems ambiguous to me. This is checking if in c there are chains through producer that reaches any group in all_consumers.
Since all consumers are consumers of c, shouldn't this actually be `!reachability.reachesAny(g, all_consumers)`? If this is already a DAG how could a consumer of c be in its producer path?

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: Checks if any is a producer of g.

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: If I want to see if a `producer` can be merged with a `consumer` I want to make sure that no consumer of producer reaches producer, or no producer of consumer reaches producer. This seems to be checking that no producer of c reaches consumers of c, or I read this wrong which could also be the case.

**Review by csarofeen — CHANGES_REQUESTED** [review]

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: Thanks for the style guideline and sorry for the confusion. I have went through the code and hopefully renamed all the confusing parts. Also removed `insert` method and used inline replacements.

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: Updated. Thanks!

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: Updated, thanks!

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: This is actually `all_consumers_of_producer`. Renamed it accordingly and sorry for the confusion.

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: Updated. Thanks!

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: Renamed `reachability` to `producer_check` to avoid confusion.

Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: Updated comment in fusion_segmenter.h to avoid confusion. 

There could be a few ways to check the merging condition. This implementation is based on Theorem 4.1 of the paper we used for segmentation (Herrmann et al.). In the case of checking if a producer and a consumer can be merged into one node. I believe based on Theorem 4.1, it is enough to check if any other consumer of the same producer also produces the consumer. e.g.

say **a** produces **b**, **c**, **d**, want to check if **a** and **b** can merge. This implementation checks 

!(**c** produces **b**) and !(**d** produces **b**) or as written in the code:
`!isConsumerOfAny(b,{c,d})`

Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: Thanks for the feedback and please let me know if there are more questions or suggestions.

**Review by csarofeen — APPROVED** [review]
Thank you, was so much easier to read after the changes. LGTM!

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: @csarofeen Wanted to see your suggestions before merging this PR. To fix some tests like layer_norm_backward, I had to run a brute force iteration before the actual merging to make sure that all scalar edges are merged into some groups. It has to be the brute force version of merging to guaranteed that no scalar edges are left behind in the end.

Also added assertion before actual merging and user can fix this assertion failure by avoiding multiple uses of a computed scalar. 

I guess in longer term if we officially not support multiple uses of scalar expressions, the initial scalar merging iterations can be much more lightweight but I'm not sure if that's reasonable request for user/parser.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: Why don't we just include all scalar computations in any graph that uses them? Scalar computations should not impact the result of `codeGenSupportedMerge`. So simply expanding to include any expression resulting in a scalar that the group uses should be fine. It may even replicate the computation of the scalar which also should be fine.

Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp]: I see. That's great idea. Thanks!

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp:1142]: @csarofeen current version inserts all the scalar producers to each group before starting merging and doesn't create groups for scalar ops anymore. Will merge this PR if it looks reasonable.

**Comment by shmsong** [comment]
Merging to unblock layer norm backward. Can address further improvements in follow ups.

**Pull request merged**
